EIPS/eip-2124.md

```diff
@@ -0,0 +1,294 @@
+
+To avoid this overhead, Ethereum needs a mechanism that can precisely identify whether a node will be useful, as early as possible. Such a mechanism requires a way to summarize chain configurations, as well as a way to disseminate said summaries in the network.
+
+This proposal focuses only on the definition of said summary - a generally useful *fork identifier* - and it's validation rules, allowing it to be embedded into arbitrary network protocols (e.g. [discovery ENRs](https://eips.ethereum.org/EIPS/eip-778) or `eth/6x` handshakes).
+
+## Abstract
+
+There are many public and private Ethereum networks, but the discovery protocol doesn't differentiate between them. The only way to check if a peer is good or bad (same chain or not), is to establish a TCP/IP connection, wrap it with RLPx cryptography, then execute an `eth` handshake. This is an extreme cost to bear if it turns out that the remote peer is on a different network and it's not even precise enough to differentiate Ethereum and Ethereum Classic. This cost is magnified for small networks, where a lot more trial and errors are needed to find good nodes.
+
+Even if the peer **is** on the same chain, during non-controversial consensus upgrades, not everybody updates their nodes in time (developer nodes, leftovers, etc). These stale nodes put a meaningless burden on the peer-to-peer network, since they just latch on to good nodes, but don't accept upgraded blocks. This causes valuable peer slots and bandwidth to be lost until the stale nodes finally update. This is a serious issue for test networks, where leftovers can linger for months.
+
+This EIP proposes a new identity scheme to both precisely and concisely summarize the chain's current status (genesis and all applied forks). The conciseness is particularly important to make the identity useful across datagram protocols too. The EIP solves a number of issues:
+
+ * If two nodes are on different networks, they should never even consider connecting.
+ * If a hard fork passes, upgraded nodes should reject non-upgraded ones, but **NOT** before.
+ * If two chains share the same genesis, but not forks (ETH / ETC), they should reject each other.
+
+This EIP does not attempt to solve the clean separation of 3-way-forks! If at the same future block number, the network splits into three (non-fork, fork-A and fork-B), separating the forkers from each another will need case-by-case special handling. Not handling this keeps the proposal pragmatic, simple and also avoids making it too easy to fork off mainnet.
+
+To keep the scope limited, this EIP only defines the identity scheme and validation rules. The same scheme and algorithm can be embedded into various networking protocols, allowing both the `eth/6x` handshake to be more precise (Ethereum vs. Ethereum Classic); as well as the discovery to be more useful (reject surely peers without ever connecting).
```

Comment:
"only defines the identity scheme..." is confusing because ENR also has 'identity schemes' which are unrelated.
Maybe best to just delete the entire paragraph TBH.

Gotta say the whole Abstract could be removed and some of the content moved into Motivation + Rationale where it really belongs.

Reply:
Yes, but the EIP *process* requires there 3 completely redundant pieces to exist, and @axic requested that I explicitly have all of them. My original EIP didn't have them. It's stupid to have them. If I've however been requested to have them, I'm not deleting them any more.

Reply:
> "only defines the identity scheme..." is confusing because ENR also has 'identity schemes' which are unrelated.

Well, this EIP is not about ENR per your request, so you should not know about ENR identity schemes at all (as a matter of fact, I've no clue what they are :D).

Reply:
> Yes, but the EIP process requires these 3 completely redundant pieces to exist, and @axic requested that I explicitly have all of them. 

Don't blame it on me, I'm just following EIP-1, hehe.

Reply:
I don't think the intention of EIP-1 is to make you repeat the same text twice with tweaked wording just so you have both sections. Maybe EIP-1 should be changed. Honestly, "simple summary" is [the definition of Abstract](https://en.wikipedia.org/wiki/Abstract_(summary)).

Reply:
@karalabe I don't understand why the fork ID is an 'identity scheme'. This sentence is the only place where 'identity' appears in the whole document. An 'identity scheme' is a way to assign identities, i.e. on a national identity card. In ENR, an identity scheme is the name of a signing function and public key crypto parameters that define how to assign a node identity.

Reply:
Well, `id` in `forkid` means identity :)

Reply:
I thought it means `identifier`.

Reply:
Hmm, yes, the forkid is an identifier, but doesn't the EIP define the rules (i.e. scheme) of how the ID is generated and regenerated over time? Maybe I have a different mental definition of "scheme". Anyway, I don't much care about the exact wording if you have a better alternative.

Reply:
Like I said I'd just kill the entire paragraph.